consensus/src/chained_bft/liveness/multi_proposer_election.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::chained_bft::{
+    common::{Author, Payload, Round},
+    consensus_types::block::Block,
+    liveness::proposer_election::ProposerElection,
+};
+
+use siphasher::sip::SipHasher13;
+use std::hash::{Hash, Hasher};
+
+/// The MultiProposer maps a round to an ordered list of authors.
+/// The primary proposer is determined by an index of hash(round) % num_proposers.
+/// The secondary proposer is determined by hash(hash(round)) % (num_proposers - 1), etc.
+/// In order to ensure the required number of proposers, a set of the proposers to choose from
+/// is updated after each hash: a chosen candidate is removed to avoid duplication.
+///
+/// Note the hash doesn't have to be cryptographic. The goal is to make sure that different
+/// combinations of consecutive leaders are going to appear with equal probability.
+
+/// While each round has more than a single valid proposer, only the primary proposer is
+/// considered for `process_proposal`.
+/// TODO: we're going to extend this functionality with an ability to use secondary proposals upon
+/// timeout.
+pub struct MultiProposer {
+    // Ordering of proposers to rotate through (all honest replicas must agree on this)
+    proposers: Vec<Author>,
+    // Number of proposers per round
+    num_proposers_per_round: usize,
+}
+
+impl MultiProposer {
+    pub fn new(proposers: Vec<Author>, num_proposers_per_round: usize) -> Self {
+        assert!(num_proposers_per_round > 0);
+        assert!(num_proposers_per_round <= proposers.len());
+        Self {
+            proposers,
+            num_proposers_per_round,
+        }
+    }
+
+    // A deterministic hashing function based on SipHash 1-3 hasher
+    pub fn hash(val: u64) -> u64 {
+        let mut hasher = SipHasher13::new();
```

Comment:
Question: Why SipHasher13 rather than SipHasher24?  It appears that the SipHasher24 has more traction form the 3rd party implementations and the research paper appears to focus more on 2-4?

Reply:
I don't understand the differences and took SipHasher13 only because it's the current default Rust hasher. My current intuition is that it doesn't really matter for as long as we have some reasonable level of randomness.
@aching if you have some context and prefer 2-4 over 1-3, I'm happy to change!

Reply:
So the research paper - https://131002.net/siphash/siphash.pdf - recommends 2-4.

"SipHash-c-d with c ≥ 2 and d ≥ 4 is expected to provide the maximum PRF
security possible (and therefore also the maximum MAC security possible) for
any function with the same key size and output size. Our fast proposal is thus SipHash-2-4".

FWIW, the default SipHasher implementation uses SipHasher24 under the covers.
The reference implementation also implements 2-4 - https://github.com/veorq/SipHash

From perusing the code, 2-4 calls compress! 8x and 1-3 calls compress! 6x.

Given this is the recommendation of the authors, the default implementation for SipHasher, and the 3rd party support is mostly for 2-4, maybe it would be best to start with 2-4.  I doubt the performance penalty will affect us and its easy to change to 1-3 if it does.